pkg/kube/krt/collection.go

```diff
@@ -357,17 +359,27 @@ func newManyCollection[I, O any](cc Collection[I], hf TransformationMulti[I, O],
 			return
 		}
 		// Now, register our handler. This will call Add() for the initial state
+		// Locking here is tricky. We want to make sure we don't get duplicate events.
+		// When we run RegisterBatch, it will trigger events for the initial state. However, other events could trigger
+		// while we are processing these.
+		// By holding the lock, we ensure we have exclusive access during this time.
+		h.recomputeMu.Lock()
 		h.eventHandlers.MarkInitialized()
-		handlerReg := c.RegisterBatch(func(events []Event[I]) {
+		handlerReg := c.RegisterBatch(func(events []Event[I], initialSync bool) {
 			if log.DebugEnabled() {
 				h.log.WithLabels("dep", "primary", "batch", len(events)).
 					Debugf("got event")
 			}
-			h.onPrimaryInputEvent(events)
+			// Lock after the initial sync only
```

Comment:
Yeah, I see what you mean about this being a little weird. Documented tho, and I can't think of a better way to solve it ATM.